Patch to Volker's extension to fix timing issues, etc.

dpvc opened the pull request:
Patch to Volker's extension to fix timing issues, prevent extra reprocess, handle configuration and menu cookies better, etc.

I will write more about this tomorrow.  I'm going to be unavailable the rest of the day, but wanted to get this started so you could look at it before Wednesday, if you wanted.  Some restructuring was needed in order to make some parts of the timing work better.  As I say, I'll explain it further soon.

zorkow commented:
Thanks. I get a couple of errors. I'll try to put some details into inline comments.

PS: I've started rewriting the Menu as discussed last week, but will hold off until this part is merged.

zorkow commented on extensions/Accessibility-Extension.js:
```diff
@@ -1,12 +1,19 @@
+  var ITEM, MENU; // filled in when MathMenu extension loads
+  
+  var BIND = (Function.prototype.bind ? function (f,t) {return f.bind(t)} :
+              function (f,t) {return function () f.apply(t,arguments)});
```
Arguments is not defined. That leads to a bug in Chrome (not FF).

dpvc replied:
arguments should *always* be defined within a function.  It refers to the arguments of the function.  Did I misspell it?   In any case, Chrome should never execute that statement, since it does have the `bind` function.

zorkow replied:
Forgot about arguments...
Anyway, tt's a scoping issue. Butting  { } around  the body of the second
function will get rid of the problem.
Why Chrome evaluates this part, I really don't know.

(I always like { } around bodies... not only because my linter tells me
to...)

dpvc replied:
D'oh.  They certainly should have been there.  I was using FF to debug, so didn't get any message.  Apologies.

zorkow commented on extensions/Accessibility-Extension.js:
```diff
@@ -44,124 +52,103 @@
-      if (result.reduce(function(x, y) {return x || y;}, false)) {
-        HUB.Queue(['Reprocess', HUB]);
+      ITEM = MathJax.Menu.ITEM;
+      MENU = MathJax.Menu.placeHolder;
```
This is undefined. Placeholders are instantiated after the CreateMenu method in the subsequent loop.
However, I assume that should be 
MENU = MathJax.Menu.menu 
anyway.

dpvc replied:
OOPS, I was doing a global find-and-replace at the last minute and that was accidentally changed.  (I had changed `placeHolder` to `menu` at one point, and then decided to put it back.)

zorkow commented on extensions/Accessibility-Extension.js:
```diff
@@ -44,124 +52,103 @@
+    HUB.Startup.signal.Post('Accessibility Loader Ready');
+  },5);   // make sure it runs before other extensions' menu hooks
+          //   even if they are loaded first
+  
```
This should go after line 145, as o/w the modules might not exist in the AddMenu method.

dpvc replied:
The `MathMenu Ready` signal should not occur until after the extensions are loaded (but perhaps that isn't true with a combined configuration file).  But if that is the case, after 145 isn't right either.  I will check into it.

zorkow commented:
Now that I got it to work, I see that you have started moving the 
So I'll stop with my rewrite entirely. Looks great with the Active at the top and the grayed out options when you switch it off.
However, when I switch off accessibility, I also can't get into the context menu anymore with Space.

dpvc commented:
> when I switch off accessibility, I also can't get into the context menu anymore with Space

That is probably because of the check I added to the `Keypress` method.  That will need to be made more sophisticated.  Sorry about that.

dpvc commented:
I added some replies to your comments, but I don't see them above.  Did you get copies, or should I add them again?

zorkow commented:
Same here. I've been frantically trying to reload the PR but they never
appeared.
Will reply by email.


On Mon, May 23, 2016 at 4:32 PM, Davide P. Cervone <notifications@github.com
> wrote:

> I added some replies to your comments, but I don't see them above. Did you
> get copies, or should I add them again?
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/mathjax/MathJax-RespEq/pull/112#issuecomment-221011650>
>

dpvc commented:
It looks like the line comments made it in after all (though I guess it took a while to register).  I have fixed the issues you identified, and have push some more commits to take care of them.  I think it is ready now.  I will outline the major changes below, and make line comments explaining some of the details in line comments.

As you know, the synchronization of timing between the extensions, and with the MathMenu extension is complicated but critical.  So a number of changes are there to coordinate that better.  You pointed out that even what I had wasn't good enough (when you suggested moving of the StartupHook).  The MathMenu extension can be loaded in a variety of ways that complicate the timing further:  it can be in a combined configuration file, which will be loaded before the extensions (and the `MathMenu Ready` will have already fired when the extension are loaded, so their MathMenu hooks will fire in the order that they are loaded, which is random), or the MathMenu could be loaded in the `extensions` array along with other extensions (rare, but allowed), in which case the `MathMenu Ready` signal will fire at a random time among the extensions (some might be laded already and others not), or MathMenu could be loaded automatically by MathJax several seconds after the initial typeset (if you don't use a combined configuration file and don't include it in the extensions explicitly).  So the point at which the menu hooks are run can vary greatly, and since the order in which the various extensions have their menus processed is important, we need to take all those possibilities into account.

I believe the code now does that.  The key is to use nested startup hooks:  one that waits for the `End Extensions` signal which will always happen after all the extensions are loaded (other than possibly the MathMenu extension, which can be loaded automatically later), and then a second one for the `MathMenu Ready` signal.  For both signals, we use handler priorities to force the order to be the one we want, regardless of the order in which extensions are loaded, and that means that the `MathMenu Ready` signals will run in the correct order, regardless of when MathMenu is loaded in relation to the extensions.  (If MathMenu comes first, we still get the proper order from the `End Extensions` order, and if it comes later, we get the correct order due to the priorities.)

The Assistive-Extension's MathMenu hook needs to run first, so that it creates the menus before the other extensions try to do so.  When the individual extensions are loaded, they look for the menu created by Assistive-Extension, and replace it if found, otherwise they add their own menu.  In the latter case, rather than always adding a rule and their own menu, by forcing the order in which the menu handlers are fired, we can replicate the menu setup used by Assistive-Extension, even if it isn't loaded.  

If the Assistive-Extension doesn't load, the Semantic-Complexity's MathMenu hook runs first and adds the first menu (and a rule).  Assistive-Explorer (which forces Semantic-MathML to load) runs next, and looks for the menu from Semantic-Complexity, which it knows will be there, and adds the Assistive menu below that.  Then the Semantic-Collapse hook runs, and again, since it forces Semantic-Complexity, it knows its menu will be there, and adds its menu in after that (which means it is between "Collapsible Math" and "Assistive", where it belongs.

Note that these can be loaded in any order, and any combination, and the result should be correctly ordered (for those menus that appear), and that this works whether Assistive-Extension is loaded or not.  (It took some effort to make that all work out.)

Since the extensions can be loaded without the Assistive-Extension, they have to be able to manage their own menus (i.e., enable and disable themselves).  Once extension is loaded, there is no real need for the Assistive-Extension to be involved with that menu any longer; it is really only there to manage the place-holder menu that will force the loading of an extension that was not active originally but is activated by the user.  So the Assistive-Extension ModuleLoader has been simplified to only deal with loading the extensions, not managing the menu after that.  Once loaded, the extension will replace the place-holder menu and it manage enable/disable actions itself (as they would if they were loaded but that page in the absence of the Assistive-Extension).

This also means that the Enable and Disable operations need to be a bit more sophisticated, since they need to make sure that the required extensions are also enabled, or dependent extensions are disabled.  So these methods have been expanded to properly signal the other extensions.  They also have to make sure the menu cookie is updated (the menu code handles updating the cookie for the extension that is endabled/disabled, but if that forces others to be enabled/disabled, those changes need to be recorded in the cookie as well).

Another timing issue is that e would like to include whatever enrichment is to be done to the math on the *first* typesetting that occurs, rather than do one typesetting without enrichment and then reprocess it again for enrichment.  This is what was happening earlier, but that leads to a poor user experience since they have to wait for the page to be typeset twice.

In order to avoid the double typeset, we need to make sure that any of the active extensions are loaded before the initial typesetting occurs.  The solution is to force the extensions to load *before* signaling that the Assistive-Extension is loaded, so that MathJax will wait for *all* the needed extensions before doing its initial typesetting pass.  That required refactoring the ModuleLoader code somewhat.

Finally, I changed the name of the first menu from "Responsive Equations" to "Collapsible Math", which is (a) shorter, and (b) more in line with it actual function: it makes the math collapsible, but it is not really "responsive" to anything ("Auto Collapse" is more the "responsive" option).

I'll make more comments in the code.

dpvc commented on extensions/Accessibility-Extension.js:
```diff
@@ -1,12 +1,19 @@
   var SETTINGS = HUB.config.menuSettings;
-  var ITEM = MathJax.Menu.ITEM;
-  var MENU = MathJax.Menu.menu;
+  var ITEM, MENU; // filled in when MathMenu extension loads
```
When not using a combined configuration file, the `MathJax.Menu` might not be defined yet (usually it is loaded after the initial typesetting).  So these need to be set by the MathMenu hook later.

dpvc commented on extensions/Accessibility-Extension.js:
```diff
@@ -1,12 +1,19 @@
+    var keys = [];
+    for (var id in obj) {if (obj.hasOwnProperty(id)) keys.push(id)}
+    return keys;
+  };
```
These allow the code to run even in IE < 9, even though selecting the accessibility options will not actually work.  We might want to do something to disable the menus entirely for early versions of IE.

dpvc commented on extensions/Accessibility-Extension.js:
```diff
@@ -28,7 +35,8 @@
-      var items = Accessibility.modules.map(function(x) {return x.placeHolder;});
+      var items = Array(this.modules.length);
+      for (var i = 0, module; module = this.modules[i]; i++) items[i] = module.placeHolder;
       var about = MENU.IndexOfId('About');
```
This is also for IE < 9

dpvc commented on extensions/Accessibility-Extension.js:
```diff
@@ -44,124 +52,104 @@
+        module.AddMethods();
+      }
+      this.AddMenu();
+    },
```
The loading of the extensions has to be done earlier, so we remove that from `Startup()`.  Since `Startup()` occurs in the MathMenu hook, we can set the `ITEM` and `MENU` values needed elsewhere.  We also create the place-holder menus here (since we had to create and register the `ModuleLoader`s before the MathMenu code may be available, and couldn't do it then).  We also add the methods (for any loaded modules).  Technically, this really isn't needed, since the menu now only loads the extension, and the extension will replace the menu, so the `ModuleLoader`s `Enable()` or `Disable()` method will not be called.  I may simplify this further.

dpvc commented on extensions/Accessibility-Extension.js:
```diff
@@ -44,124 +52,104 @@
       }
+      return (extensions.length ? HUB.Startup.loadArray(extensions) : null);
     }
   };
```
This is used to load any extensions that are requested by the menu settings (i.e., ones that were checked by the user earlier or by the default configuration set up by the author).  This has to be done before the Assistive-Extension calls `loadComplete` so that we are sure all the needed extensions have been loaded before the initial typeset pass.  (See the call to `LoadExtensions` at the end of the file.)

dpvc commented on extensions/Accessibility-Extension.js:
```diff
@@ -44,124 +52,104 @@
     placeHolder: null,
-    activeBox: null,
-    subMenu: null,
+    submenu: false,
```
This is now a signal for whether the extension's version of the menu will be a submenu (like `Assistive`) and should have an `Activate` entry, as we discussed in the hangout last week.

dpvc commented on extensions/Accessibility-Extension.js:
```diff
@@ -44,124 +52,104 @@
+    Init: function(option, name, module, extension, submenu) {
       this.option = option;
-      this.name = name;
+      this.name = [name.replace(/ /g,''),name];
```
In order to be properly localized, we need to use `[id,name]` for the menu name, where the `id` is used for identifying the menu, and `name` is the string that will be localized (and changes from language to language).  Currently, the `id` is just the `name` with spaces removed, but in the future, the pair could be passed as the `name` argument directly.

dpvc commented on extensions/Accessibility-Extension.js:
```diff
@@ -44,124 +52,104 @@
-      this.placeHolder = ITEM.CHECKBOX(
-        this.name, Accessibility.MakeOption(this.option), {action: helper});
-      this.activeBox = ITEM.CHECKBOX(
-        'Active', Accessibility.MakeOption(this.option), {action: helper});
```
We can't create the menu at this point because the `ModuleLoader` may be created before the MathMenu extension is loaded.  So we put the creation of place-holder menus into the `Startup()` method, which happens when MathMenu loads.

dpvc commented on extensions/Accessibility-Extension.js:
```diff
@@ -44,124 +52,104 @@
-      return flag;
+      this.submenu = (submenu || false);
     },
-    Switch: function() {
```
We don't need a `Switch()` method any longer, since the only action the place-holder menu does is load the extension.  After that, the extension replaces the place-holder, and the place-holder is no longer used.  We never go back.

dpvc commented on extensions/Accessibility-Extension.js:
```diff
@@ -44,124 +52,104 @@
+          this.name, Accessibility.MakeOption(this.option), {action: load}
+        );
       }
     },
```
The `CreateMenu()` method is called during `Startup()` to create place-holder menus that load the extensions.  The proper type of menu (`CHECKBOX` or `SUBMENU`) is created based on the needs of the `ModuleLoader`.

dpvc commented on extensions/Accessibility-Extension.js:
```diff
@@ -44,124 +52,104 @@
-          that.Disable = ext.Disable;
-          if (that.onSwitch) that.Enable(true);
-        }
-      });
```
The modified version of this is below (I guess I inserted routines above this, and the `git diff` didn't get the connection between the two versions).

dpvc commented on extensions/Accessibility-Extension.js:
```diff
@@ -44,124 +52,104 @@
-      });
+    Load: function() {
+      HUB.Queue(["Require",MathJax.Ajax,this.module,["AddMethods",this,true]]);
     },
```
The load action is to use `MathJax.Ajax.Require()` to load the function, calling `this.AddMethods(true)` when that is done.

dpvc commented on extensions/Accessibility-Extension.js:
```diff
@@ -44,124 +52,104 @@
+    Load: function() {
+      HUB.Queue(["Require",MathJax.Ajax,this.module,["AddMethods",this,true]]);
     },
-    AddActiveBox: function() {
```
This is no longer needed.

dpvc commented on extensions/Accessibility-Extension.js:
```diff
@@ -44,124 +52,104 @@
+        if (menu) {
+          this.Enable(true,true);
+          MathJax.Menu.saveCookie();
         }
```
Here we get the Enable/Disable methods, and if called with `true` (i.e., from the load action of a place-holder menu), we enable the extension and save the menu cookie (since the enabling may have enabled other extensions as well whose cookie state needs to be saved).

dpvc commented on extensions/Accessibility-Extension.js:
```diff
@@ -44,124 +52,104 @@
-      if (index !== null) {
-        MENU.items[index] = this.subMenu;
       }
     }
```
We don't need `AddPlaceHolder()` or `AddSubMenu()` any longer, as the extensions manage their own menus (and rep ace the place-holders) when they are loaded.

dpvc commented on extensions/Accessibility-Extension.js:
```diff
@@ -44,124 +52,104 @@
+      'explorer', 'Accessibility', '[RespEq]/Assistive-Explore.js',
+      'Assistive', true
+    )
+  );
```
We create and register the loaders immediately, since we need to know what extensions to load before `loadComplete()` is called (so they are loaded before the first typeset pass, avoiding an extra reprocess step and user frustration).

dpvc commented on extensions/Accessibility-Extension.js:
```diff
@@ -44,124 +52,104 @@
+    )
+  );
 
+  Accessibility.AddDefaults();
```
We can now set the `menuSettings` for the registered extensions.

dpvc commented on extensions/Accessibility-Extension.js:
```diff
@@ -44,124 +52,104 @@
+    ["LoadExtensions",Accessibility],
+    ["loadComplete",MathJax.Ajax,"[RespEq]/Accessibility-Extension.js"]
+  );
 
```
We load any extensions that have been selected by the user or the page author, and then signal that our loading is complete (so the MathJax startup process can continue on; the `End Extensions` signal should follow this).

dpvc commented on extensions/Accessibility-Extension.js:
```diff
@@ -44,124 +52,104 @@
-        'Explorer Ready', 'Assistive'));
-    MathJax.Extension.Accessibility.Startup();
-  MathJax.Hub.Startup.signal.Post('Accessibility Loader Ready');
-}));
```
All of this has to be done earlier, since this queued action occurs *after* the initial typeset pass.  Since we need to load extensions *before* that typesetting, the registration of modules needs to be done earlier, as discussed above, and in the main comments.

dpvc commented on extensions/Assistive-Explore.js:
```diff
@@ -4,6 +4,7 @@
 MathJax.Hub.Register.StartupHook('Sre Ready', function() {
   var FALSE, KEY;
   var SETTINGS = MathJax.Hub.config.menuSettings;
+  var COOKIE = {};   // replaced when menu is available
```
We need to maintain some menu cookies by hand, since enabling one extension may cause others to be enabled, and the menu code only save the cookie for the item that was selected, not the other ones.

dpvc commented on extensions/Assistive-Explore.js:
```diff
@@ -12,6 +13,7 @@ MathJax.Hub.Register.StartupHook('Sre Ready', function() {
 
   var Assistive = MathJax.Extension.Assistive = {
     version: '1.0',
+    dependants: [],            // the extensions that depend on this one
```
If other extensions rely on this one, and this is disabled, we need to disable the other ones.  So we need to keep a list of the dependents to disable when we are disabled.

dpvc commented on extensions/Assistive-Explore.js:
```diff
@@ -92,16 +94,46 @@ MathJax.Hub.Register.StartupHook('Sre Ready', function() {
+    Enable: function(update,menu) {
+      SETTINGS.explorer = true;
+      if (menu) COOKIE.explorer = true;
+      MathJax.Extension.SemanticComplexity.Enable(false,menu);
```
Here, we set the `menuSettings` for the explorer to `true`, and if this is a menu action, we set the cookie as well.  Then we signal the Semantic-Complexity extension (on which we rely) to enable itself as well.

dpvc commented on extensions/Assistive-Explore.js:
```diff
@@ -92,16 +94,46 @@ MathJax.Hub.Register.StartupHook('Sre Ready', function() {
+      if (!this.hook) {
+        this.hook = MathJax.Hub.Register.MessageHook( 
+          'New Math', ['Register', this.Explorer]);
       }
```
Then we enable our submenu entries (disable = false), and if there is not already a global hook, we set one up.

[The reason you couldn't stop the hook was because the `Enable()` was being called more than once, and that means two message hooks were registered, with the second causing you to loose the first value of `this.hook`.  The `Disable()` method then only removed one of the two hooks, so the second continued to be called when the `New Math` message occurred.]

dpvc commented on extensions/Assistive-Explore.js:
```diff
@@ -92,16 +94,46 @@ MathJax.Hub.Register.StartupHook('Sre Ready', function() {
+      for (var i = this.dependants.length-1; i >= 0; i--) {
+        var dependant = this.dependants[i];
+        if (dependant.Disable) dependant.Disable(false,menu);
+      }
```
Here we unset the `menuSettings` value, and unset the cookie if this is a menu action.  We disable the submenu items (disabled = true), and remove the message hook, if any.  Finally, we tell any dependent extensions (none at this point, but maybe there will be in the future) to disable themselves.

dpvc commented on extensions/Assistive-Explore.js:
```diff
@@ -92,16 +94,46 @@ MathJax.Hub.Register.StartupHook('Sre Ready', function() {
+          for (var i = 2, item; item = items[i]; i++) item.disabled = state;
+        }
+      }
+    },
```
This function goes through the submenu and disables/enables all the entries (other than the `Active` item at the top).

dpvc commented on extensions/Assistive-Explore.js:
```diff
@@ -92,16 +94,46 @@ MathJax.Hub.Register.StartupHook('Sre Ready', function() {
+    //
+    Dependant: function (extension) {
+      this.dependants.push(extension);
     }
```
This allows an extension to register the fact that it depends on Assistive-Explore.

dpvc commented on extensions/Assistive-Explore.js:
```diff
@@ -238,6 +270,7 @@ MathJax.Hub.Register.StartupHook('Sre Ready', function() {
     // Registers new Maths and adds a key event if it is enriched.
     //
     Register: function(msg) {
+      if (!Assistive.hook) return;
```
The presence of the `hook` indicates that the Assistive-Explore extension is enabled.  So if it is not there, don't register any new math.  (This is just a precaution, since we should now be adding and removing only a single hook, but it would have taken care of the duplicate hook problem in the original code.)

dpvc commented on extensions/Assistive-Explore.js:
```diff
@@ -314,11 +347,13 @@ MathJax.Hub.Register.StartupHook('Sre Ready', function() {
       math.addEventListener(
           Explorer.focusinEvent,
           function(event) {
+            if (!Assistive.hook) return;
```
We make the event handlers check the `hook` (i.e., when the extension is enabled) so we don't have to remove all the hooks in the `Disabled()` method (i.e., we don't have to do a `Reprocess()` call simply to remove the handlers).

dpvc commented on extensions/Assistive-Explore.js:
```diff
@@ -610,100 +645,104 @@ MathJax.Hub.Register.StartupHook('Sre Ready', function() {
       Assistive.addDefaults();
-      Assistive.Enable();
+      this.createMenu();
+    },
```
The `responsive` settings are now handled in the Semantic-Complexity's `Enable()` method, and we don't want to force it to true in any case, as the user's menu setting may require it to be false.  Similarly, we don't want to force ourselves to be active if the user's menu cookie says otherwise.  (Enabling and disabling are handled below.)

But we do want to tell Semantic-Complexity that we are dependent on it (so if it becomes disabled, it will tell us to disable as well).

dpvc commented on extensions/Assistive-Explore.js:
```diff
@@ -610,100 +645,104 @@ MathJax.Hub.Register.StartupHook('Sre Ready', function() {
+      this.createMenu();
+    },
+    
+    createMenu: function () {
```
The menu code had to be moved to the `Startup()` routine in order to handle the timing right, but now that it is wrapped in the handler for `End Extensions`, that wasn't necessary.  Argh!  I will change that, now that I see it.

dpvc commented on extensions/Assistive-Explore.js:
```diff
@@ -610,100 +645,104 @@ MathJax.Hub.Register.StartupHook('Sre Ready', function() {
+    createMenu: function () {
+      MathJax.Hub.Register.StartupHook('End Extensions', function () {
+        Assistive[SETTINGS.explorer === false ? "Disable" : "Enable"]();
+        MathJax.Hub.Startup.signal.Post('Explorer Ready');
```
We do the Enable/Disable call here so that if the Assistive-Extension has been loaded, it has the chance to set the `SETTINGS` value first.  (If we load Assistive-Explore by itself, that means the page author wants it active by default; but Assistive-Extension turns everything off by default, so we need to wait for Assistive-Explore to run, if it is loaded, before we can tell what the default should be.)  We put off the ready signal until we are actually enabled or disabled.

dpvc commented on extensions/Assistive-Explore.js:
```diff
@@ -610,100 +645,104 @@ MathJax.Hub.Register.StartupHook('Sre Ready', function() {
+        MathJax.Hub.Startup.signal.Post('Explorer Ready');
+        MathJax.Hub.Register.StartupHook('MathMenu Ready', function () {
+          COOKIE = MathJax.Menu.cookie;
+          if (SETTINGS.explorer == null) SETTINGS.explorer = true;
```
We save the menu cookie object for later use, and set the `SETTINGS` value if it isn't already.

dpvc commented on extensions/Assistive-Explore.js:
```diff
@@ -610,100 +645,104 @@ MathJax.Hub.Register.StartupHook('Sre Ready', function() {
+          var Switch = function(menu) {
+            Assistive[SETTINGS.explorer ? "Enable" : "Disable"](true,true);
+            MathJax.Menu.saveCookie();
+          };
```
The menu action is just to call the `Enable()` or `Disable()` function (informing it to reprocess the page, if needed, and that this action is being done from a menu).  Then we save the menu cookie, in case other extensions were enabled/disabled along the way.

dpvc commented on extensions/Assistive-Explore.js:
```diff
@@ -610,100 +645,104 @@ MathJax.Hub.Register.StartupHook('Sre Ready', function() {
+          } else {
+            index = MENU.IndexOfId('CollapsibleMath');
+            MENU.items.splice(index+1,0,accessibilityMenu);
+          }
```
Here we find the location of the menu.  It should either replace the existing accessibility menu (created by Assistive-Extension) or should follow the "Collapsible Math" item.

dpvc commented on extensions/Assistive-Explore.js:
```diff
@@ -610,100 +645,104 @@ MathJax.Hub.Register.StartupHook('Sre Ready', function() {
+            index = MENU.IndexOfId('CollapsibleMath');
+            MENU.items.splice(index+1,0,accessibilityMenu);
+          }
+          Assistive[SETTINGS.explorer ? "Enable" : "Disable"]();
```
I think this one can be removed, since the one at 656 should already have done it.

dpvc commented on extensions/Assistive-Explore.js:
```diff
@@ -610,100 +645,104 @@ MathJax.Hub.Register.StartupHook('Sre Ready', function() {
+          }
+          Assistive[SETTINGS.explorer ? "Enable" : "Disable"]();
+        },20);  // Between Semantic-Complexity and Semantic-Collapse
+      },20);
```
We use priority 20, which is in between the Semantic-Complexity's priority of 15 and Semantic-Collapse's priority of 25.  That means the "Collapsable Math" menu will be there (from Semantic-Complexity), but when Semantic-Collapse adds "Auto Collapse" right after "Collapsible Math", it will move our menu down, and get the right order.

dpvc commented on extensions/Assistive-Explore.js:
```diff
@@ -610,100 +645,104 @@ MathJax.Hub.Register.StartupHook('Sre Ready', function() {
-  MathJax.Extension.SemanticComplexity.Enable();
   MathJax.Extension.Assistive.Explorer.Startup();
-  MathJax.Hub.Startup.signal.Post('Explorer Ready');
   MathJax.Ajax.loadComplete('[RespEq]/Assistive-Explore.js');
```
W don't do the `Enable()` call here (it is done at line 656, and there it is conditional on the state of the menu selections).  We also don't post the ready signal until 657, as discussed there.

dpvc commented on extensions/Semantic-Collapse.js:
```diff
@@ -3,27 +3,63 @@
+        if (dependant.Disable) dependant.Disable(false,menu);
+      }
+      if (update) HUB.Queue(["Rerender",HUB]);
     },
```
These are similar to the ones in Assistive-Explore, except that when we are asked to update, we do that.

dpvc commented on extensions/Semantic-Collapse.js:
```diff
@@ -3,27 +3,63 @@
+      //  Inform SemanticComplexity that we are a dependant
+      //
+      var Complexity = MathJax.Extension.SemanticComplexity;
+      if (Complexity) Complexity.Dependant(this);
```
We register as dependent on Semantic-Complexity.

dpvc commented on extensions/Semantic-Collapse.js:
```diff
@@ -36,6 +72,35 @@
+          }
+        },25);  // after Assistive-Explore
+      },25);
+    },
```
This is similar to Assistive-Explore.  We use `End Extensions` to coordinate among the extensions (using the priorities of the handlers), and handle the configuration from the menu settings or the menu settings from the configuration (depending on whether there was an explicit menu choice).  Then create the menu and place it as described above.

dpvc commented on extensions/Semantic-Collapse.js:
```diff
@@ -251,51 +317,9 @@
-      }
-      MathJax.Menu.config.settings['collapse'] ?
-        Collapse.Enable() : Collapse.Disable();
-    },
```
This routine was never used, even in the original version (which is whey the `resizeHandler` changes that we made last week didn't have any effect).  So we don't need this.

dpvc commented on extensions/Semantic-Collapse.js:
```diff
@@ -251,51 +317,9 @@
-    Remove: function() {
-      Collapse.Disable();
-      MathJax.Extension.SemanticComplexity.Disable();
     }
```
This is now handled by `Disbaled()` itself.

dpvc commented on extensions/Semantic-Complexity.js:
```diff
@@ -8,6 +8,9 @@
   
+  var SETTINGS = HUB.config.menuSettings;
+  var COOKIE = {};   // replaced when menu is available
+  
```
The changes here are similar to the ones for Semantic-Collapse above.

dpvc commented on extensions/Semantic-MathML.js:
```diff
@@ -4,6 +4,7 @@
 MathJax.Extension.SemanticMathML = {
   version: "1.0",
   config: MathJax.Hub.CombineConfig("SemanticMathML",{disabled: false}),
+  dependants: [],     // the extensions that depend on this one
```
These changes are similar to the two extensions above.

dpvc commented on extensions/Semantic-MathML.js:
```diff
@@ -71,6 +89,10 @@ MathJax.Callback.Queue(
+  //  Load toMathML extension (if it isn't already)(
+  //
+  ["Require",MathJax.Ajax,"[MathJax]/extensions/toMathML.js"],
+  //
```
Since this extension relies on toMathML, I force it to load as well.

dpvc commented on extensions/Accessibility-Extension.js:
```diff
@@ -44,124 +52,104 @@
+      Accessibility.Startup();
+      HUB.Startup.signal.Post('Accessibility Loader Ready');
+    },5);   // run before other extensions' menu hooks even if they are loaded first
+  },5);
```
We use the `End Extensions` signal to make sure that the order of menu operations is correct (as discussed in the main comments), since the `MathMenu Ready` signal can occur in essentially random places.  We use the priority (5 in this case, which is less than the default priority of 10) to schedule the Assistive-Extension menus to be made before any of the loaded extensions, which use higher priorities (i.e., later execution).

dpvc commented on extensions/Accessibility-Extension.js:
```diff
@@ -44,124 +52,104 @@
-          this.AddMethods();
-          this.AddActiveBox();
-        }
-      }
```
We no longer need this `Load()` method (its replacement appears below).

dpvc commented:
OK, I have made the changes that I mentioned in the line notes.  I think this is ready at this point.  Please check it out and see if you find anything else.

dpvc commented:
OK, I've pushed some changes that should complete the extension as per our discussion this morning.  It is not as tested as I would like, but I am out of time for this afternoon.  Will do more testing later, but wanted you all to have access to the changes sooner rather than later.
